subprojects/core/src/main/java/org/gradle/api/internal/provider/AbstractMappingProvider.java

```diff
@@ -60,4 +62,13 @@ public OUT getOrNull() {
     public String toString() {
         return "transform(" + provider + ")";
     }
+
+    @Override
+    public void visitDependencies(TaskDependencyResolveContext context) {
```

Comment:
I don't think we want to do this, as we don't know that the result of the mapping is built by the task or not. When it's not built by the task (e.g. it's only the locations) then we shouldn't carry any task information or possibly should fail when this is used as a task input.

We also don't know whether the mapping needs the result of the task or not (in which case we shouldn't run the mapping until after the task has executed), or in the case of files, whether the consumer needs only the location (don't need to run the tasks that produce the files, but _do_ run the tasks that determine the location, if any) or whether the consumer needs the location and the contents (do run the tasks that produce the files).

I think we need some better ways of describing what the mapping does and what the consumer needs instead of this change.

Reply:
Generally related: I was thinking about this aspect of Provider and friends recently, and was wondering about using different types to model file locations and contents. To what extent would clearly decoupling these aspects make things better? I don't have anything more concrete than that question to add.